typescript/cli/src/rebalancer/monitor/Monitor.ts

```diff
@@ -26,51 +24,74 @@ export class Monitor implements IMonitor {
-    const mailboxes = objMap(addresses, (_, { mailbox }) => ({ mailbox }));
-    const provider = new MultiProtocolProvider(objMerge(metadata, mailboxes));
-    const warpCoreConfig = await this.registry.getWarpRoute(this.warpRouteId);
-    const warpCore = WarpCore.FromConfig(provider, warpCoreConfig);
-
-    // Start the interval used to poll collateral balances
-    this.interval = setInterval(async () => {
-      const event: MonitorEvent = {
-        balances: [],
-      };
-
-      for (const token of warpCore.tokens) {
-        // Ignore non-collateralized tokens given that we only care about collateral balances
-        if (!token.isCollateralized()) {
-          continue;
-        }
+    try {
+      this.isMonitorRunning = true;
+
+      // Build the WarpCore from the registry
+      const metadata = await this.registry.getMetadata();
+      const addresses = await this.registry.getAddresses();
+
+      // The Sealevel warp adapters require the Mailbox address, so we
+      // get mailboxes for all chains and merge them with the chain metadata.
+      const mailboxes = objMap(addresses, (_, { mailbox }) => ({ mailbox }));
+      const provider = new MultiProtocolProvider(objMerge(metadata, mailboxes));
+      const warpCoreConfig = await this.registry.getWarpRoute(this.warpRouteId);
+      const warpCore = WarpCore.FromConfig(provider, warpCoreConfig);
+
+      // this can be considered the starting point of the monitor
+      this.emitter.emit('start');
+
+      while (this.isMonitorRunning) {
+        const event: MonitorEvent = {
+          balances: [],
+        };
+
+        for (const token of warpCore.tokens) {
+          // Ignore non-collateralized tokens given that we only care about collateral balances
+          if (!token.isCollateralized()) {
+            continue;
+          }
 
-        const adapter = token.getHypAdapter(warpCore.multiProvider);
+          const adapter = token.getHypAdapter(warpCore.multiProvider);
 
-        // Get the bridged supply of the collateral token to obtain how much collateral is available
-        const bridgedSupply = await adapter.getBridgedSupply();
+          // Get the bridged supply of the collateral token to obtain how much collateral is available
+          const bridgedSupply = await adapter.getBridgedSupply();
 
-        event.balances.push({
-          chain: token.chainName,
-          owner: token.addressOrDenom,
-          token: token.collateralAddressOrDenom!,
-          value: bridgedSupply!,
-        });
+          event.balances.push({
+            chain: token.chainName,
+            owner: token.addressOrDenom,
+            token: token.collateralAddressOrDenom!,
+            value: bridgedSupply!,
+          });
+        }
+
+        // Emit the event containing the collateral balances
+        this.emitter.emit('collateralbalances', event);
+
+        // Wait for the specified check frequency before the next iteration
+        await sleep(this.checkFrequency);
```

Comment:
This should be outside the catch if we want the monitor to keep working as intended if a poll fails